lib/improver/utilities/statistical_operations.py

```diff
@@ -0,0 +1,208 @@
+        if output_name is not None:
+            self.output_name = output_name
+        else:
+            self.output_name = "probability_of_{}".format(
+                percentiles_cube.name())
+
+    def __repr__(self):
+        """Represent the configured plugin instance as a string."""
+        result = ('<ProbabilitiesFromPercentiles2D: percentiles_cube: {}, '
+                  'output_name: {}'.format(self.percentiles_cube,
+                                           self.output_name))
+        return result
+
+    def create_probability_cube(self, cube):
+        """
+        Create a 2-dimensional probability cube in which to store the
+        calculated probabilities.
+
+        Args:
+            cube (iris.cube.Cube):
+                Template for the output probability cube.
+        Returns:
+            probability_cube (iris.cube.Cube):
+                A new 2-dimensional probability cube with suitable metadata.
+        """
+        cube_format = next(cube.slices([cube.coord(axis='y'),
+                                        cube.coord(axis='x')]))
+        probabilities = cube_format.copy(data=np.full(cube_format.shape,
+                                                      np.nan, dtype=float))
+        probabilities.units = 1
+        probabilities.rename(self.output_name)
+        return probabilities
+
+    def percentile_interpolation(self, threshold_cube, percentiles_cube):
+        """
+        Perform the interpolation between 2-dimensional percentile fields to
+        construct the probability field for a given set of thresholds.
+
+        Args:
+            threshold_cube (iris.cube.Cube):
+                A 2-dimensional cube of "threshold" values for which it is
+                desired to obtain probability values from the percentiled
+                reference cube.
+            percentiles_cube (iris.cube.Cube):
+                A cube of 2-dimensional fields on several different percentile
+                levels.
+        Returns:
+            probabilities (iris.cube.Cube):
+                A 2-dimensional cube of probabilities obtained by interpolating
+                between percentile values.
+        """
+        probabilities = self.create_probability_cube(percentiles_cube)
+
+        percentile_coordinate = find_percentile_coordinate(percentiles_cube)
+        percentiles = percentile_coordinate.points
+
+        threshold_heights = threshold_cube.data.flatten()
+        pdata = np.full(threshold_heights.shape, np.nan, dtype=float)
+
+        percentile_point_heights_list = []
+        for cube in percentiles_cube.slices(percentile_coordinate):
+            percentile_point_heights_list.append(cube.data.flatten())
+
+        # loop over spatial points in 2D field
+        for index, (percentile_heights, threshold_height) in enumerate(
+                zip(percentile_point_heights_list, threshold_heights)):
+            pdata[index] = 0.01*np.interp(threshold_height, percentile_heights,
```

Comment:
@cgsandford Depending upon what assumptions are made, this section of code where the cube data is flattened, so that the spatial points can be looped over, could be replaced by:
```python
pdata = np.interp(threshold_cube.data, np.unique(percentiles_cube.data), percentiles)
```
This would mean that there was a single interpolation for all grid points, rather than interpolating separately for each grid point.